### PR TITLE
pkp/pkp-lib#7318 add export button for publication stats

### PR DESCRIPTION
**Pull request by bozana**
s. https://github.com/pkp/pkp-lib/issues/7318

This PR contains a few fixes and refactoring, and the export possibility on the manager statistics > articles/monographs/preprints page.

### CODE REVIEW

File: classes/context/Context.php [hunk 1]
```diff
@@ -546,6 +546,26 @@ public function isInstitutionStatsEnabled(Site $site): bool
         }
         return (bool) $enableInstitutionUsageStats;
     }
+
+    /**
+    * What Geo data to track for usage statistics.
+    * Consider context setting only if the site setting is enabled and context setting considers less Geo data than site setting.
+    *
+    * @return ?string Possible return values: null, disabled, country, coutnry+region, country+region+city
```
Comment (NateWr) [inline classes/context/Context.php:554]: `country+region`

File: classes/context/Context.php [hunk 2]
```diff
@@ -546,6 +546,26 @@ public function isInstitutionStatsEnabled(Site $site): bool
...
+    */
+    public function getGeoStatsSetting(Site $site): ?string
```
Comment (NateWr) [inline classes/context/Context.php:556]: Let's call this `getEnableGeoUsageStats` so it is the same as the setting name.

File: classes/context/Context.php (hunk 1 above)
Comment (NateWr) [reply classes/context/Context.php:554]: Is there any difference in how the system work between `null` and `disabled`? I wonder if we should set a value by default and enforce either `disabled` or one of the three options? Seems like we shouldn't need `null` and `disabled`.

File: classes/context/Context.php [hunk 3]
```diff
@@ -546,6 +546,26 @@ public function isInstitutionStatsEnabled(Site $site): bool
...
+    {
+        $enableGeoUsageStats = $site->getData('enableGeoUsageStats');
+        if ($enableGeoUsageStats !== null) {
+            if (($enableGeoUsageStats != 'disabled') &&
+                ($this->getData('enableGeoUsageStats') !== null) &&
+                ($this->getData('enableGeoUsageStats') != $site->getData('enableGeoUsageStats')) &&
+                    str_starts_with($enableGeoUsageStats, $this->getData('enableGeoUsageStats'))) {
+                $enableGeoUsageStats = $this->getData('enableGeoUsageStats');
+            }
+        }
```
Comment (NateWr) [inline classes/context/Context.php:566]: I admit it took me a while to understand the logic here. What about being a bit more explicit about the different states that can happen?

```php
$siteSetting = $site->getData('enableGeoUsageStats');
if ($siteSetting !== null || $siteSetting === 'disabled') {
    return $siteSetting;
}
$contextSetting = $this->getData('enableGeoUsageStats');
if (str_starts_with($siteSetting, $contextSetting)) {
    return $contextSetting;
}
return $siteSetting;
```

File: classes/core/APIResponse.php [hunk 4]
```diff
@@ -41,7 +41,7 @@ public function withCSV(array $rows, array $columns, ?int $maxRows): self
         $this->getBody()->write($csvData);
         $this->withStatus(200);
         if (isset($maxRows)) {
-            $this->withHeader('X-Total-Count', $maxRows);
+            return $this->withHeader('X-Total-Count', $maxRows)->withHeader('Content-Type', self::RESPONSE_CSV);
```
Comment (NateWr) [inline classes/core/APIResponse.php:44]: Let's make `$maxRows` required and always return `X-Total-Count`. That will make the API responses more predictable and simplify our code.

File: classes/statistics/PKPStatisticsHelper.php [hunk 5]
```diff
@@ -152,23 +152,14 @@ public static function hashIp(string $ip, string $salt): ?string
...
-                    if ($enableGeoUsageStats == 'country+region+city') {
-                        $city = $geoIPArray['city'];
-                    }
+        $enableGeoUsageStats = $context->getGeoStatsSetting($site);
+        if ($enableGeoUsageStats != 'disabled') {
+            $geoIPArray = $this->getLocation($ip, $hashedIp, $flushCache);
+            $country = $geoIPArray['country'];
+            if ($enableGeoUsageStats == 'country+region+city' || $enableGeoUsageStats == 'country+region') {
+                $region = $geoIPArray['region'];
+                if ($enableGeoUsageStats == 'country+region+city') {
```
Comment (NateWr) [inline classes/statistics/PKPStatisticsHelper.php:161]: Doesn't need to be done now, but can we turn these setting strings into constants? They could go in `PKPStatisticsHelper` or `Context`.

File: templates/stats/publications.tpl [hunk 6]
```diff
@@ -172,6 +172,82 @@
...
 								</a>
 							</div>
+							<pkp-button
+								ref="downloadReportModalButton"
+								@click="$modal.show('downloadReport')"
+							>
+								{translate key="common.downloadReport"}
+							</pkp-button>
+							<modal
+								close-label="Close"
```
Comment (NateWr) [inline templates/stats/publications.tpl:182]: This needs to be localized. You can use `close-label="{translate key="common.close"}"` in template files.

File: classes/context/Context.php (hunk 1 above)
Comment (bozana) [reply classes/context/Context.php:554]: There is no difference how the system behaves.
I have an extra issue regarding this, that I would like to take a look at with you: https://github.com/pkp/pkp-lib/issues/7976.

**Pull request merged**